cse modifications

smichr opened the pull request:
- closes #10228
- closes #11230

smichr committed:
cse modifications

- closes #10228
- closes #11230

smichr commented on sympy/simplify/cse_main.py:
```diff
@@ -212,42 +212,120 @@ def _find_opts(expr):
+            return Func(*args, evaluate=False)
+
+        def as_dict(e):
+            # return a well formed unevaluated function from the args
```
correct docstring

smichr commented on sympy/simplify/cse_main.py:
```diff
@@ -212,42 +212,120 @@ def _find_opts(expr):
+                # don't allow a sign to change
+                if com_dict[di] > func_dicts[k][di]:
+                    return
+            # remove it
```
remove it as many times as possible -- see test failure where `x+y` is extracted once from `4x+4y`

asmeurer commented on sympy/simplify/cse_main.py:
```diff
@@ -212,42 +212,120 @@ def _find_opts(expr):
+            # SHARES Func
+            args = []
+            for a in c:
+                if not (
```
Maybe more readable:

```
if not all((dx[a].is_Integer, dx[a] > 0,
    dy[a].is_Integer, dy[a] > 0)):

smichr replied:
those lines are no longer present

asmeurer commented on sympy/simplify/cse_main.py:
```diff
@@ -212,42 +212,120 @@ def _find_opts(expr):
 
-        func_args = [set(e.args) for e in funcs]
+        # ----------------- helpers ---------------------------
+        def cfunc(dx, dy, c):
```
Is it possible to move these functions out to the module level? They are impossible to test independently if they are inside another function.

asmeurer replied:
It's also complicated enough that it could be worth adding a small docstring with a doctest (even if it's a private function).

smichr replied:
I moved out the one more general routine (`pairwise_most_common`) but the other ones are helpers that are written as functions to simplify the function in which they are contained. I provided comments at the start and throughout those functions. See if it looks any better now.

smichr committed:
MatExpr.args_cnc doesn't report 1 as coeff

MatExpr was reporting the args_cnc as ([1], [...]) for a Mul expression
but Expr does not report the [1] -- it gives []. By correcting this,
some code is simplified in cse.

Also, since there is a doctest showing how a coefficient of a MatExpr
can have a non-commutative entity, the comment regarding that was
removed.

smichr commented:
ping @asmeurer

moorepants commented:
FYI, this introduces a 4.29x decrease in CSE computation speed in our benchmarks:

http://www.moorepants.info/misc/sympy-asv/#large_exprs.TimeLargeExpressionOperations.time_cse?commits=9670ff89

asmeurer commented:
That's not good. @smichr any thoughts on how to improve that?

smichr commented:
> thoughts on how to improve that

not right off...I would hope that the resulting expressions, having made better extractions, will pay back the investment in code that uses them. But I'll be looking in the weeks to come for ways to make the cse better. I'll also be working on getting the non-commutative extraction incorporated.

smichr commented:
@moorepants , I made some mods to cse that I pushed to my cse2 branch. Could you try your benchmark (or send me the benchmark) on that branch?

asmeurer commented:
I believe [this](https://github.com/sympy/sympy_benchmarks/blob/master/benchmarks/large_exprs.py#L71) is the benchmark. You can run the benchmarks in that repo with the run_benchmarks.sh script (it just runs airspeed velocity). Running the full airspeed velocity will probably take a long time, since it runs the benchmarks on every commit over a range.

smichr commented:
@asmeurer , thanks for that reference. I ran the following in master and the cse2 branch and it looks like things have improved considerably (at least for that expression):

```python
import sympy as sm
txt=('''
MutableDenseMatrix([[Add(Mul(Symbol('l0'), Symbol('m1'), Pow(Function('u1')(Symbol('t')),
Integer(2)), cos(Function('q1')(Symbol('t')))), Mul(Symbol('l0'), 
...
''')
t = sm.symbols('t')
funcs = [s(t) for s in sm.symbols('q:6, u:6', cls=sm.Function)]
syms = sm.symbols('x:{}'.format(len(funcs)))
subs = dict(zip(funcs, syms))
func_matrix = sympify(txt)
sym_matrix = func_matrix.subs(subs)
long_expr = 0
for expr in sym_matrix[:]:
    long_expr += expr

from time import time
for i in range(4):
    t=time();ans=cse(long_expr);time()-t, count_ops(ans)

master
(0.93, 306)
(0.63, 306)
(0.63, 306)
(0.62, 306)

cse2
(0.33, 297)
(0.28, 297)
(0.28, 297)
(0.28, 297)
(0.28, 297)
```

asmeurer commented:
How about before this PR?

smichr commented:
Ah, good point:

```
(0.16, 272)
(0.14, 272)
(0.14, 272)
(0.14, 272)
```

Summarizing the replacements, 

```
pre old pr found expressions like this:
count  expr pattern
----- ------------
20      a*b
29      a*b*c
5       a**2
12      a*(b + c)

while this new pr found this:
count  expr pattern
----- ------------
30      a*b
21      a*b*c
4       a**2
4       a + b
4       a*b*c*d
8       a*b*c*(d + e)
```

smichr commented:
@asmeurer , this is not going to be a quick fix for me. I am in favor of reverting the change until a better solution is found. My current solution for the long_expr, tested as

```python
import sympy as sm
code = \
r'''begin 666 -
M'XL( ,80YE<"_^6:S8K;0!"$[_T4N4F&!*+IG@<(A$ .AD".BP_[8Q:#UR9K
MF21O']MAC7_&,].CZM%ACUJT4U*I>^I3RS3=]O</R_G7^6HSG][WKXL_[=W=
MEZ>G=KI=MC__OCRLEVVS_-Q,/GYX.WKI]D<_UK_;;]O58[]8K]IFN_O;\?2^
MF>Q.^+[JY\_SU];M#Q[7FY.S?UV>O3_E6I".BJZ2XE&0LP0)J"@FMT@116]O
M:G<F>'B*=*'H5(I.IQAZC E!&J8H>L4B08H^1K"B2YK*NMY@G:)D*6H%*:+H
M313/!/GZ%B][0U2*HE,,W:+H>B.I*$E%K[I%?R4XF>TNZQA8;__[J3O1?3Z]
MB%"49>YJZL7_[W=6JS/DTBD61U:7[IO!Z16*EF 5;!:KJ$S6;GBXD(N57'*E
MZUM2[M118@A%'=E9@'. K"R0 45 52U0.I#?!_X=]H$SZP/.*P+&] &C+$#V
M@:T%2@<HVP)@'W!F']2Q(.) ""@Q%DA>$0BF#P1E@0?V@:T%2@=B?2!6?> S
M^\!#+(B^%QRH/P/Z*4:+F8&? ;==DLMO9JEZ=;&\=-^4#$"RN1R'(+%*QO!(
M-Q!*Z;P>H!84.$ C6' CBL:RH"2*8!9X71'0>!8H':#;L] 0F.=$D:N(9 Z-
M9%V*R\G. IP#9&6!'U $5-6"0:\F70K,,44@F7T@$ NDO @"7$YV%N <()P%
M8M8'/L\"#RF"-)>KIO$NB<Y<#K>AW==J==\4? (*3#!0W,@%Q,"C<Z-#<B/(
M@LK<Z)#<6-&".MP(XZ0S!\@$&9P--]I88,.-SH0;<:AH8,&@3SL.P8U<D1L9
MS8T.PHU<CQL9SHT.P8U<D1NY/C=R<B@JY63'Z:&HZ- NTLF0CS')H:B,/13E
MP(_6H!84.& .MUWV]SG8>^^P(K .=4:&>D4+ZH0Z+,2B<(L)=;8)=1L+;$)=
M90$C0ETJAKK4#W5)3E2\+G8EO]+*OG^G'I@?89R0/7;$31J'66"]X^9;P)@B
=(!."QFXWM_I 1NH#&?X:-9G-)O0/<W;DR-$T
end
'''
def tug(string):
    """Return text of a uu encoded gzipped text."""
    import StringIO, uu, gzip
    f1 = StringIO.StringIO()
    f2 = StringIO.StringIO()
    f2.write(string)
    f2.seek(0)
    uu.decode(f2, f1)
    f2.seek(0)
    f2.flush()
    f1.seek(0)
    text = gzip.GzipFile(fileobj=f1, mode='rb')
    msg = text.read()
    [f.close() for f in [text, f1, f2]]
    return msg

code = tug(code)
import hashlib
assert hashlib.md5(str(code)).hexdigest() == \
'dda9cb8b6b35ba975bc61b890a242281'
func_matrix = sympify(code)
t = sm.symbols('t')
funcs = [s(t) for s in sm.symbols('q:6, u:6', cls=sm.Function)]
syms = sm.symbols('x:{}'.format(len(funcs)))
subs = dict(zip(funcs, syms))
sym_matrix = func_matrix.subs(subs)
long_expr = 0
for expr in sym_matrix[:]:
    long_expr += expr

for i in range(4):
    t=time();ans=cse(long_expr);S(time()-t).n(2), count_ops(ans) + len(ans[0
```

gives this:

```
(0.63, 321)
(0.37, 321)
(0.33, 321)
(0.33, 321)
```

smichr commented:
One important thing that I've learned to watch for is that when extracting factors, you have to watch out for cases where the bases are something like x and y from expressions such as `(x*y**2, x*y**3*z, x*y**2*z)`. If you only extract `x0 = x*y` you will get `(x0*y, x0*y**2*z, x0*y*z)` from which `x1 = x0*y` can be extracted to give `(x1, x1*y*z, x1*z)` from which, finally, `x2 = x1*z` can be extracted to give `(x1, x2*y, x2)`. It would have been better to start the extraction with `x0 = x*y**2` to give `(x0, x0*y*z, x0*z)` and then `x1 = x0*z` to give `(x0, x1*y, x1)`. In either case, the moral of the story is that the substituted patterns will have residual extractions that can be made and this is perhaps the reason for the observation that @moorpants made in #11428.

asmeurer commented:
This also seems to have led to a test failure https://github.com/sympy/sympy/issues/11684

isuruf commented:
This leads to a bug as well. #12070
Should we revert the PR, until it can be fixed?

asmeurer commented:
We may want to do that, or try a fix. I haven't heard from @smichr in a while.

bjodah commented:
The algorithmic complexity of this new `cse` functions looks (empirically) to be exponential:
https://github.com/sympy/sympy_benchmarks/pull/38
Already 2 expressions with less than 20 terms takes essentially "forever" to process - we had to limit our benchmark to 11 variables.
